x/ccv/provider/keeper/misbehaviour.go

```diff
@@ -0,0 +1,150 @@
+	}
+
+	// jail and tombstone the byzantine validators
+	for _, v := range evidence.ByzantineValidators {
+		// convert consumer consensus address
+		consuAddr := sdk.ConsAddress(v.Address.Bytes())
+		provAddr := k.GetProviderAddrFromConsumerAddr(ctx, misbehaviour.Header1.Header.ChainID, types.ConsumerConsAddress{consuAddr})
+		k.stakingKeeper.ValidatorByConsAddr(ctx, consuAddr)
+		val, ok := k.stakingKeeper.GetValidatorByConsAddr(ctx, provAddr.Address)
+
+		if !ok || val.IsUnbonded() {
+			logger.Error("validator not found or is unbonded", provAddr.String())
+			continue
+		}
+
+		// jail validator if not already
+		if !val.IsJailed() {
+			k.stakingKeeper.Jail(ctx, provAddr.Address)
+		}
+
+		// tombstone validator if not already
+		if !k.slashingKeeper.IsTombstoned(ctx, provAddr.Address) {
+			k.slashingKeeper.Tombstone(ctx, provAddr.Address)
+		}
+
+		// update jail time to end after double sign jail duration
+		k.slashingKeeper.JailUntil(ctx, provAddr.Address, evidencetypes.DoubleSignJailEndTime)
+	}
+
+	logger.Info(
+		"confirmed equivocation",
+		"byzantine validators", evidence.ByzantineValidators,
+	)
+
+	return nil
+}
+
+// ConstructLigthClientEvidence constructs and returns a CometBFT Ligth Client Attack(LCA) evidence struct
+// from the given misbehaviour
+func (k Keeper) ConstructLigthClientEvidence(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) (*tmtypes.LightClientAttackEvidence, error) {
+
+	// construct the trusted and conflicetd ligth blocks
+	trusted, err := headerToLightBlock(*misbehaviour.Header1)
+	if err != nil {
+		return nil, err
+	}
+	conflicted, err := headerToLightBlock(*misbehaviour.Header2)
+	if err != nil {
+		return nil, err
+	}
+
+	// get common header using the IBC misbehaviour
+	commonHeight, commonTs, commonValset, err := k.GetTrustedInfoFromMisbehaviour(ctx, misbehaviour)
+	if err != nil {
+		return nil, err
+	}
+
+	// construct the LCA evidence by copying the CometBFT constructor
+	// see newLightClientAttackEvidence() in tendermint/light/detector.go
+	ev := tmtypes.LightClientAttackEvidence{
+		ConflictingBlock: conflicted,
+	}
+
+	if ev.ConflictingHeaderIsInvalid(trusted.Header) {
+		ev.CommonHeight = commonHeight
+		ev.Timestamp = commonTs
+		ev.TotalVotingPower = commonValset.TotalVotingPower()
+	} else {
+		ev.CommonHeight = trusted.Header.Height
+		ev.Timestamp = trusted.Header.Time
+		ev.TotalVotingPower = trusted.ValidatorSet.TotalVotingPower()
+	}
+
+	ev.ByzantineValidators = ev.GetByzantineValidators(commonValset, trusted.SignedHeader)
+
+	return &ev, nil
+}
+
+// GetCommonFromMisbehaviour checks whether the given ibc misbehaviour's headers share common trusted height
+// and that a consensus state exists for this height. In this case, it returns the associated trusted height, timestamp and valset.
+func (k Keeper) GetTrustedInfoFromMisbehaviour(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) (int64, time.Time, *tmtypes.ValidatorSet, error) {
+
+	// a common trusted height is required
```

Comment:
Why is this a requirement? This is not a requirement in ibc-go when freezing the client.

Reply:
It's necessary to construct a `tmtypes.LightClientAttackEvidence` which requires the common header of the `misbehaviour`'s conflicting headers.

The Hermes relayer follows this requirement to construct `MsgSubmitMisbehaviour` msgs.

Reply:
correct.